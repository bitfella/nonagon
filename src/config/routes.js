export const routes = {
  about: '/about',
  logout: '/logout',
  callback: '/callback'
};
