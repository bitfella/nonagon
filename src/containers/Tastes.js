import React, { Component } from 'react';
import { connect } from 'react-redux';
import { lowerCase } from 'lodash-es';
import styled, { css } from 'styled-components/macro';
import PropTypes from 'prop-types';
import { getTopTracks } from 'actions/tastes';
import {
  getToken,
  getTastesTimeRange,
  playlistLoading,
  createPlaylistError,
  addTrackToPlaylistError,
  recommendationsLoading,
  recommendationsError,
  tastesError,
  tastesLoading
} from 'selectors';
import {
  appTimeRange,
  spotifyWebApiTimeRanges as timeRanges
} from 'config/spotify';
import Error from 'components/Error';
import Spinner from 'components/Spinner';
import TopTracks from 'containers/TopTracks';

const TastesWrapper = styled.div`
  padding-bottom: 5rem;
`;

const RangeSwitch = styled.nav`
  display: flex;
  padding: 1rem;
`;

const RangeButtonActive = css`
  background-color: ${props => props.theme.colors.c};
  cursor: default;
`;

const RangeButton = styled.button`
  flex-grow: 1;
  padding: 0.5rem 0.25rem;
  background-color: ${props => props.theme.colors.f};
  border: 1px solid ${props => props.theme.colors.c};
  color: ${props => props.theme.colors.g};
  font-size: 0.6875rem;
  font-weight: 600;
  text-transform: uppercase;
  cursor: pointer;

  &:first-child {
    border-top-left-radius: 1rem;
    border-bottom-left-radius: 1rem;
  }

  &:not(:first-child) {
    border-left: none;
  }

  &:last-child {
    border-top-right-radius: 1rem;
    border-bottom-right-radius: 1rem;
  }

  ${props => props.range === props.activeRange && RangeButtonActive}
`;

export class Tastes extends Component {
  static propTypes = {
    token: PropTypes.string,
    timeRange: PropTypes.string,
    getTracks: PropTypes.func,
    playlistLoading: PropTypes.bool,
    createPlaylistError: PropTypes.object,
    addTrackToPlaylistError: PropTypes.object,
    recommendationsLoading: PropTypes.bool,
    recommendationsError: PropTypes.object,
    tastesError: PropTypes.object,
    tastesLoading: PropTypes.bool
  };

  getTopTracks = (token, range) => {
    localStorage.setItem(appTimeRange, range);
    this.props.getTracks(token, range);
  };

  render() {
    const {
      token,
      timeRange,
      playlistLoading,
      createPlaylistError,
      addTrackToPlaylistError,
      recommendationsLoading,
      recommendationsError,
      tastesLoading,
      tastesError
    } = this.props;

    if (tastesLoading || playlistLoading || recommendationsLoading)
      return <Spinner />;
    if (tastesError) return <Error data={tastesError} />;
    if (recommendationsError) return <Error data={recommendationsError} />;
    if (createPlaylistError) return <Error data={createPlaylistError} />;
    if (addTrackToPlaylistError)
      return <Error data={addTrackToPlaylistError} />;

    return (
      <TastesWrapper>
        <RangeSwitch>
          {timeRanges.map((range, index) => {
            return (
              <RangeButton
                activeRange={timeRange}
                range={range}
                key={index}
                onClick={() => this.getTopTracks(token, range)}
              >
                {lowerCase(range)}
              </RangeButton>
            );
          })}
        </RangeSwitch>
        <TopTracks />
      </TastesWrapper>
    );
  }
}

const mapStateToProps = state => ({
  token: getToken(state),
  timeRange: getTastesTimeRange(state),
  playlistLoading: playlistLoading(state),
  createPlaylistError: createPlaylistError(state),
  addTrackToPlaylistError: addTrackToPlaylistError(state),
  recommendationsLoading: recommendationsLoading(state),
  recommendationsError: recommendationsError(state),
  tastesError: tastesError(state),
  tastesLoading: tastesLoading(state)
});

const mapDispatchToProps = dispatch => ({
  getTracks: (token, range) => dispatch(getTopTracks(token, range))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Tastes);
