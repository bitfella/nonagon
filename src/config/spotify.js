import { routes } from 'config/routes';

export const appAccessToken = 'accessToken';
export const appTimeRange = 'timeRange';
export const appRedirectURI = window.location.origin + routes.callback;
export const spotifyWebApiBasePath = 'https://api.spotify.com/v1';
export const spotifyWebApiProfileEndpoint = `${spotifyWebApiBasePath}/me`;
export const spotifyWebApiUserTopTracksEndpoint = `${spotifyWebApiBasePath}/me/top/tracks`;
export const spotifyWebApiGetRecommendationsEndpoint = `${spotifyWebApiBasePath}/recommendations`;
export const spotifyWebApiCreatePlaylistEndpoint = `${spotifyWebApiBasePath}/users/___USER_ID___/playlists`;
export const spotifyWebApiAddTrackToPlaylistEndpoint = `${spotifyWebApiBasePath}/playlists/___PLAYLIST_ID___/tracks`;
export const spotifyWebApiGetPlaylistImagesEndpoint = `${spotifyWebApiBasePath}/playlists/___PLAYLIST_ID___/images`;
export const spotifyWebApiTimeRanges = [
  'short_term',
  'medium_term',
  'long_term'
];
export const spotifyWebApiInitialRefinementValue = [0.0, 1.0];
export const spotifyWebApiScopes = [
  'user-top-read',
  'playlist-modify-public',
  'playlist-modify-private'
];
export const spotifyWebApiScopesGuideURI =
  'https://developer.spotify.com/documentation/general/guides/scopes/';
export const spotifyAuthURI = 'https://accounts.spotify.com/authorize';
export const spotifyLoginURI =
  `${spotifyAuthURI}` +
  `?client_id=${process.env.REACT_APP_SPOTIFY_CLIENT_ID}` +
  `&response_type=token` +
  `&scope=${encodeURIComponent(spotifyWebApiScopes.join(' '))}` +
  `&redirect_uri=${appRedirectURI}`;
export const spotifyLogoutURI = 'https://accounts.spotify.com/logout';
