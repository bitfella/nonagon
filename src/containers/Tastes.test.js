import React from 'react';
import { shallow } from 'enzyme';
import { Tastes } from './Tastes';
import { spotifyWebApiTimeRanges as timeRanges } from 'config/spotify';

const props = {
  token: 'ACCESS_TOKEN',
  timeRange: timeRanges[0]
};

describe('Tastes', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Tastes {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
