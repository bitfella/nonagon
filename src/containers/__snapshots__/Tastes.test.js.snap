// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tastes renders correctly 1`] = `
<Tastes__TastesWrapper>
  <Tastes__RangeSwitch>
    <Tastes__RangeButton
      activeRange="short_term"
      key="0"
      onClick={[Function]}
      range="short_term"
    >
      short term
    </Tastes__RangeButton>
    <Tastes__RangeButton
      activeRange="short_term"
      key="1"
      onClick={[Function]}
      range="medium_term"
    >
      medium term
    </Tastes__RangeButton>
    <Tastes__RangeButton
      activeRange="short_term"
      key="2"
      onClick={[Function]}
      range="long_term"
    >
      long term
    </Tastes__RangeButton>
  </Tastes__RangeSwitch>
  <Connect(TopTracks) />
</Tastes__TastesWrapper>
`;
